README.md

# TBT


[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1241518.svg)](https://doi.org/10.5281/zenodo.1241518)
[![EEGLABplugin](https://img.shields.io/badge/eeglab-TBT-9cf)](https://sccn.ucsd.edu/eeglab/plugin_uploader/plugin_list_all.php)


This EEGLAB plugin allows for the automatic rejection and interpolation of channels on an epoch-by-epoch basis.

It also comes with an additional method for rejecting epochs - [max-minus-min threshold](./eegmaxmin/README.md).

You can reference the plugin and its documentation as follows:

- Ben-Shachar, M. S. (2018). TBT: Reject and Interpolate channels on a trial-by-trial basis. Zenodo. [https://doi.org/10.5281/zenodo.1241518](https://doi.org/10.5281/zenodo.1241518)

See ***Scripting*** below for a citation example.

## Downloading

-   You can download TBT through EEGLAB's data-processing extension GUI (*File &gt; Manage EEGLAB Extensions &gt; Filter by artifact*).
-   Or as a `.zip` file from the GitHub page [(TBT)](https://github.com/mattansb/TBT/releases).

## List of included functions

-   `eegplugin_TBT` - EEGLAB plugin function.
-   `tbt_bcr2` - epoch-by-epoch channel interpolation, based on any rejection method used by EEGLAB, or a manual cell-array listing which channels to interpolate in which epochs.
-   `tbt_bool2cell` - converts a boolean channel-by-trial matrix to a tbt ready cell-list (see bellow).
-   `tbt_cell2bool` - converts a tbt ready cell-list a boolean channel-by-trial matrix (see bellow).
-   `pop_TBT` - call from EEGLAB menu of `tbt_bcr`. If no parameters are specified, pops a GUI window.

## TBT GUI

Use the menu *Tools &gt; Epoch by Epoch Rejection/Interpolation*, or type into the command line:

```Matlab
[EEG, com, badlist] = pop_TBT(EEG); % pop-up interactive window mode
```

![](img/TBT_eg.png)

You will be asked to select a rejection method, and set its parameters, and also to set the following:
1. **The maximum percent of bad trials per channel.** If a channel is bad on more then this percent of the trials, the channel will be removed across the whole data-set.
2. **The maximum number of bad channels per trial.** If a trial has more than this number of bad channels, the trial will be removed.
3. Whether to plot the marked channels and trials before rejecting and interpolating the marked channels.

If you select `Plot before executing`, a pop-up window will appear, allowing for two kinds of plots:

- ***A scrolling EEG plot***, with bad channels marked in red, and bad trials marked in yellow:  

![](img/tbt_plot_eeg.png)

- ***A matrix plot***:
  - Red horizontal bars mark channels that will be completely removed,  
  - Red vertical bars mark trials that will be completely removed,  
  - Colored dots mark which channels that will be removed and interpolated on a trial-by-trial basis.  

![](img/tbt_plot_matrix.png)


## Scripting

Scripting takes to general following form:

```Matlab
% Use some rejection method:
EEG = pop_eegmaxmin(EEG);

% Send the 'rejE' matrix to pop_TBT:
my_bads = EEG.reject.rejmaxminE;
EEG = pop_TBT(EEG,my_bads,10,0.3);

% to get more info, type 'help pop_TBT' in the command line.
```

Scripting gives two major additional not available in the gui:

1.  Supplying a cell-list for rejection
2.  interpolating additional missing channels (thus making TBT an optional last step in pre-processing)

### Supplying a cell-list

A cell list can be manually created to mark bad channels in specific trials. For example, if we wish to remove E12 and E45 from the 1st epoch, and E22 from epochs 13 and 28, we would create a cell list to be used this list as input for `pop_TBT`:

```Matlab
my_bads = {...
  1,{'E12','E45'};...
  [13 28],{'E22'};...
  }
  
EEG = pop_TBT(EEG,my_bads,10,0.3);
                  
```

This method can also be combined with other `rejE` methods using the `tbt_cell2bool` function:

```Matlab
% Specify cell-list
my_bads = {...
  1,{'E12','E45'};...
  [13 28],{'E22'};...
  }

% transforms the cell-list into a 'rejE'-like matrix.
my_bads = tbt_cell2bool(my_bads,EEG);

% combine with automatic method:
EEG = pop_eegmaxmin(EEG);


my_bads = my_bads | EEG.reject.rejmaxminE;
EEG = pop_TBT(EEG,my_bads,10,0.3);
                  
```
The `tbt_bool2cell` function is the reverse of `tbt_cell2bool`, converting a boolean `rejE` matrix to a tbt-ready cell-list. For example:

```Matlab

EEG = pop_eegmaxmin(EEG);

tbt_bool2cell(EEG.reject.rejmaxminE, EEG)

>> ans =
>> 
>>   70ª2 cell array
>> 
>>     [ 2]    {'E64'	'E90'}
>>     [ 3]    {'E63'	'E64'	'E68'	'E90'	'E99'}
>>     [ 5]    {'E68'	'E73'}
>>      ...

```

### Interpolating all missing channels

By default, trial-by-trial interpolation interpolates *only* the channels that are marked on a single-trial basis. i.e., channels marked as bad across the whole data-set will not be re-added by interpolation. If you wish to add them back (or any other channel that may have been removed in any previous processing step), channel locations can be added to `pop_TBT`:

```Matlab
% To add back all channels from the input EEG data-set:
EEG = pop_eegmaxmin(EEG);

my_bads = EEG.reject.rejmaxminE;

EEG = pop_TBT(EEG,my_bads,10,0.3,[],EEG.chanlocs); % or any other chanloc struct
                  
```

This makes TBT an ideal 'last step' in preprocessing - providing a clean data-set with all missing channels interpolated.

### Citation Example

This is a citation example for the following code:

```Matlab
EEG = pop_eegmaxmin(EEG, [],[], 75, [], 1, 0);

EEG = pop_TBT(EEG, EEG.reject.rejmaxminE , 10, 0.15, 1);
```

of the max-min method + TBT:

> Epoched data were subjected to an automated bad-channel and artifact detection using EEGPLAB's *TBT* plugin (Ben-Shachar, 2018): within each epoch, channels that exceeded a differential average amplitude of 75μV were marked for rejection. Channels that were marked as bad on more then 15/% of all epochs were excluded. Epochs having more than 10 bad channels were excluded. Epochs with less than 10 bad channels were included, while replacing the bad-channel data with spherical interpolation of the neighboring channel values.

Author
------

-   **Mattan S. Ben-Shachar** \[aut, cre\] \<mattanshm at post.bgu.ac.il\>


NEWS.md
# TBT 2.6.2

- `tbt_bcr2()` will now let you know if you have channels in your data that are *not* (for some reason) in a passed `chanloc`.

# TBT 2.6.1

- `tbt_bool2cell` can now return a concise list, making the TBT process faster.
- When closing the plotting dialog via the `X` button, the function does NOT run.

# TBT 2.6.0

- Works better with the new `eeglab2019`.
- Improved plotting menu.


# TBT 2.5.0

- `pop_TBT` now returns the epochs in the correct order (and also runs a lot faster) (previously, due to the function of `pop_mergeset`, the output EEG's trials were not necessarily in the same order as the input EEG - but the `urevents` could have been used as usual).
- When plotting in `pop_TBT`, a second array-plot is returned, allowing for a 'glance' at the rejected trials / channels.
- Fixed bug in `pop_eegmaxmin` where default value for `winSize` did not make sense.
- Fixed bug in `pop_TBT` where the returned `EEG` dataset would sometimes consist of a single concatenated trial, and not epoched data (caused by `eeglab`'s `pop_mergeset`).


tbt/pop_TBT.m
% pop_TBT() - Rejects and iterpolates channels on a epoch by epoch basis.
%
% Usage:
%   >>  EEG = pop_TBT(EEG,bads,badsegs,badchans,plot_bads,chanlocs);
%
%   only interpolate channels according to `bads`:
%   >>  EEG = pop_TBT(EEG,bads,EEG.nbchan,1,plot_bads,chanlocs); 
%
%   pop-up interative window mode:
%   >> [EEG, com, badlist]   = pop_TBT(EEG);
%
%
% When called from the eeglab GUI, the pop-up window will require the
% following:
%   Rejection Methods:
%       Rejection Method    - select which of the 5 standard eeglab
%                             rejection method (+1) to use.
%       Method argument     - Arguments to me passed to the rejection
%                             method function (the text below specifiys
%                             what needs to be defined).
%   Rejection Criteria:
%       Max % bad epochs per channel    - see bellow.
%       Max N bad channel per epoch     - see bellow.
%       Plot before executing           - if checked, will plot before
%                                         executing. 
%
% Inputs:
%   EEG         - input dataset
%   bads        - a boolean 2-d matrix (channels * trials) specifying "bad
%                 channels" for each epoch. OR
%                 a x-by-2 cell list, with the first column specifying the
%                 epochs, and the second column specifying the bad channels
%                 in those epochs. e.g.: {1,{'E12','E45'};[13 28],{'E22'}}
%                 will remove E12 and E45 from the 1st epoch, and E22 from
%                 epochs 13 and 28.
%   badsegs     - Number of max bad channels per epoch. If an epoch has
%                 more than this number of bad channels, the epoch is
%                 removed.
%   badchans    - Proportion (e.g., 0.3) of max bad epochs per channel. If
%                 a channel was found to be bad on more than this percent
%                 of trials, it is removed.
%   plot_bads   - [0|1] plot before executing. When plotting, will also ask
%                 to confirm. If no plotting, will execute immediately.
%   chanlocs    - [optional] a chanlocs struct (e.g., EEG.chanlocs). If
%                 provided, missing channels will be interpolated according
%                 to this struct, and not the input EEG. NOTE that if not
%                 provided, channel that have been rejected across the
%                 dataset (according to the badchans critirion) will not be
%                 interpolated back in.
%    
% Outputs:
%   EEG     - output dataset.
%   badlist - number of bad channel and epochs removed
%
% Copyright (C) 2017  Mattan S. Ben-Shachar
%
% This program is free software; you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation; either version 2 of the License, or
% (at your option) any later version.
%
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.
%
% You should have received a copy of the GNU General Public License
% along with this program; if not, write to the Free Software
% Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

function [EEG, com, badlist] = pop_TBT(EEG,bads,badsegs,badchans,plot_bads,chanlocs)

com = '';

if nargin < 1
   help pop_TBT;
   return;
elseif nargin < 4
    
    methods = ['Abnormal values|'...
        'Abnormal trends|'...
        'Improbable data|'...
        'Abnormal distributions|'...
        'Abnormal spectra|'...
        'Max-Min Threshold'];

    methods2 = {...
        ['-10 , 10 ,' num2str(EEG.xmin) ' , ' num2str(EEG.xmax)],...
        'lowthresh, upthresh, starttime, endtime',...
        'pop_eegthresh';...
        ...
        [num2str(EEG.pnts) ' , 0.5 , 0.3'],...
        'winsize, maxslope, minR',...
        'pop_rejtrend';...
        ...
        '3 , 3',...
        'locthresh, globthresh',...
        'pop_jointprob';...
        ...
        '3 , 3',...
        'locthresh, globthresh',...
        'pop_rejkurt';...
        ...
        ['''method'' , ''FFT'' , ''threshold'' , [-30 , 30] ,''freqlimits'' , [15 , 30]'],...
        ' ',...
        'pop_rejspec';...
        ...
        ['[1:' num2str(EEG.nbchan) '],[' num2str([EEG.xmin EEG.xmax]*1000) '],100,' num2str([EEG.xmax - EEG.xmin]*1000) ',1,0'],...
        'chanRange,timeRange,minmaxThresh,winSize,stepSize,maW',...
        'pop_eegmaxmin';...
        };

    badTrial = num2str(max(ceil((EEG.nbchan-64)/6.5),4));
    
    [~, ~, ~, res] = inputgui( 'geometry', {[2.5] [0.6 1.4 0.5] [0.6 1.9] [0.6 1.9] [2.5] [2.5] [1 0.3 1.2] [1 0.3 0.2 1]}, 'geomvert', [1 1 1 1 0.5 1 1 1],...
        'title', 'pop_TBT() - Epoch by Epoch Rejection / Interpolation',...
        'uilist', { ...
        {'Style', 'text', 'string', 'Rejection Methods', 'fontweight', 'bold'  }...
        {'Style', 'text', 'string', 'Rejection Method' }...
        {'Style', 'popupmenu', 'string', methods, 'callback' @set_method 'tag' 'method' } ...
        {'Style', 'pushbutton', 'string', 'help', 'callback', @get_help}...
        {'Style', 'text', 'string', 'Method argument' }...
        {'Style', 'edit', 'string', methods2{1,1} 'tag' 'options' } ...
        {}...
        {'Style', 'text', 'string', methods2{1,2} 'tag' 'hint' } ...
        {}...
        {'Style', 'text', 'string', 'Rejection Criteria', 'fontweight', 'bold'  }...
        {'Style', 'text', 'string', 'Max % bad epochs per channel' }...
        {'Style', 'edit', 'string', '0.3' 'tag' 'badChan' }...
        {}...
        {'Style', 'text', 'string', 'Max N bad channel per epoch' }...
        {'Style', 'edit', 'string', badTrial 'tag' 'badTrial' }...
        {}...
        {'Style', 'checkbox', 'string' 'Plot before executing' 'value' 1 'tag' 'plot' }...
        } );
    
    try
    switch res.method
        case 1
            comrej  = ['EEG = pop_eegthresh(EEG, 1, 1:EEG.nbchan,' res.options ', 1, 0);'];
            bads    = 'EEG.reject.rejthreshE';
        case 2
            comrej  = ['[EEG, comrej] = pop_rejtrend(EEG, 1, 1:EEG.nbchan,' res.options ', 1, 0,0);'];
            bads    = 'EEG.reject.rejconstE';
        case 3
            comrej  = ['[EEG, ~,~,~,comrej] = pop_jointprob(EEG, 1, 1:EEG.nbchan,' res.options ', 1, 0, 0);'];
            bads    = 'EEG.reject.rejjpE';
        case 4
            comrej  = ['[EEG, ~,~,~,comrej] = pop_rejkurt(EEG, 1, 1:EEG.nbchan,' res.options ', 1, 0, 0);'];
            bads    = 'EEG.reject.rejkurtE';
        case 5
            comrej  = ['[EEG, ~, comrej]    = pop_rejspec(EEG, 1,' res.options ');'];
            bads    = 'EEG.reject.rejfreqE';
        case 6
            comrej  = ['[EEG, comrej]    = pop_eegmaxmin(EEG,' res.options ');'];
            bads    = 'EEG.reject.rejmaxminE';
    end    
    
    comtbt = ['EEG = pop_TBT(EEG,' bads ',' res.badTrial ',' res.badChan ',' num2str(res.plot) ');'];
    com = [comrej comtbt];
    eval([comrej]);
    eval([comtbt]);
    end
else
    if ~exist('plot_bads','var'), plot_bads = 1; end
    if exist('chanlocs','var')
        [EEG, nbadchan, nbadtrial] = tbt_bcr2(EEG,bads,badsegs,badchans,plot_bads,chanlocs);
    else
        [EEG, nbadchan, nbadtrial] = tbt_bcr2(EEG,bads,badsegs,badchans,plot_bads);
    end
    
    badlist.nbadchan  = nbadchan;
    badlist.nbadtrial = nbadtrial;
end
    
    % Set default method's text options and hint
    % ------------------------------------------
    function set_method(x,y)
        M_ind = get(gcbo, 'value');
        set(findobj(gcbf, 'tag', 'options'), 'string',methods2{M_ind,1});
        set(findobj(gcbf, 'tag', 'hint'), 'string',methods2{M_ind,2});
    end

    % Pop method's help
    % -----------------
    function get_help(x,y)
        M_ind = get(findobj(gcbf, 'tag', 'method'), 'value');
        eval(['pophelp(methods2{M_ind,3})'])
    end

end

